app/components/OutcomeBar.tsx

```diff
@@ -1,64 +1,136 @@
-'use client';
-
-import { FixedProductMarketMaker } from '@/queries/omen';
 import { cx } from 'class-variance-authority';
+import { useQuery } from '@tanstack/react-query';
+import request from 'graphql-request';
+
+import { BlockDataType } from '@/queries/xdai';
+import {
+  FixedProductMarketMaker,
+  FpmmTrade_OrderBy,
+  getMarketTrades,
+} from '@/queries/omen';
+import { OMEN_SUBGRAPH_URL, XDAI_BLOCKS_SUBGRAPH_URL } from '@/constants';
 import { Outcome } from '@/entities';
 
 interface OutcomeBarProps {
   market: FixedProductMarketMaker;
 }
 
+type OutcomeTokenMarginalPricesResponse = {
+  [key: string]: {
+    outcomeTokenMarginalPrices: string[];
+  };
+};
+
 export const OutcomeBar = ({ market }: OutcomeBarProps) => {
+  const { id } = market;
+
+  const { data: trades } = useQuery({
+    queryKey: ['getMarketTrades', id],
+    queryFn: async () =>
+      getMarketTrades({
+        fpmm: id,
+        orderBy: FpmmTrade_OrderBy.CreationTimestamp,
+      }),
+  });
+
+  const { data: lastTradeMarginalPrices } = useQuery({
+    queryKey: ['getLastTradeMarginalPrices', id],
+    queryFn: async (): Promise<string[] | undefined> => {
+      if (!trades) return;
+
+      const lastTradeTimestamp =
+        trades.fpmmTrades[trades.fpmmTrades.length - 1]?.creationTimestamp;
+
+      const blockNumber = await request<BlockDataType>(
+        XDAI_BLOCKS_SUBGRAPH_URL,
+        `query blockNumberByTimestamp {
+          _${lastTradeTimestamp}: blocks(where: {timestamp: ${lastTradeTimestamp}}, first: 1) {
+              number
+            }
+        }`
+      );
```

Comment:
Let's try to have something like this. The query can even go outside. `BlockDataType` will be different.

```suggestion
      const blockNumber = await request<BlockDataType>(
        XDAI_BLOCKS_SUBGRAPH_URL,
        "query blockNumberByTimestamp({ $lastTradeTimestamp: Int }) {
            blocks(where: {timestamp: $lastTradeTimestamp}, first: 1) {
              number
            }
        }",
        { lastTradeTimestamp }
      );
```

Reply:
Pointed! Working on it, thanks!